WordPress/gutenberg
# Refactor performance tests setup

## What?

Refactor the performance testing environment setup to improve developer experience both locally and in CI.

TL;DR:
- Cut setup time by 50% through the following improvements:
  - a. Utilize the local code for the reference environment and test source.
  - b. Run branch environment setups in parallel.
- Limit disk space usage by reusing the setup base folder.
- Allow rerunning tests with existing setup to save time during local development.
- Always use the current root branch as the test source to prevent issues with env misalignments.
- Use the latest stable WordPress version by default instead of the bleeding edge for consistent results.

Read on for more details.

### ⚡︎ Reduce disk space usage to a fixed 3GB

While debugging the script, after a couple of reruns, I started running out of disk space. It turned out that the script is creating a new test environment (repo clones + wp-env docker images) with each run without clearing it when finished. Currently, the initial perf script run takes ~8GB (6GB for 3 repo clones + 2GB for wp-env docker images), and each next takes another ~6GB for new clones.

Furthermore, one of the comparison branches is **always** the branch that we're currently checked out at (true for each CI job), but for some reason, it's being cloned twice from the origin: one for the test environment and one for the test source. This takes not only extra space but also time.

The refactored approach includes two changes to address the above:
1. Use a fixed name for the temporary base folder (`wp-gutenberg-performance-tests`),
2. Use the current branch for the test environment and source.

This setup requires 3GB for a typical job where we compare two branches, e.g., the current one vs. `trunk`. 


### ⚡︎ Use the current root branch code directly

Let's say your current branch is `try/some-stuff` and you'd like to compare its performance against `trunk`. To do that, you need to run the following from the gutenberg root:

```
 bin/plugin/cli.js perf try/some-stuff trunk
```

Currently, this will not use your local code. Instead, it will clone the branch to be used as the env source. It will also create another copy of that branch to be used as the source of the tests. In reality, we always use the current branch as a reference (as each workflow perf comparison job does), so using what's already available makes more sense. For the CI, it means that it will only need to build the current branch, allowing us to drop the clones and extra (uncached) `npm` installs. Locally, it means we can test against the currently developed code, which isn't currently possible. 


### ⚡︎ Use the current root branch as a tests source

In addition to the above, the current branch should always be used as the tests source because they should always be coupled with the workflow and the CLI runner script. Currently, the tests branch is either the first one passed for comparison or the one specified by the `--test-branch` option. This approach can cause issues, as exemplified by the [currently failing release performance tests](https://github.com/WordPress/gutenberg/actions/runs/4680828980) where the branch used as the tests source (`wp/6.1`) is not aligned with workflow and CLI source (`release/15.6`).

Because of the above, keeping the `--test-branch` option didn't make sense, so I've removed it.


### ⚡︎ Enable rerunning the tests using the existing environment (locally)

Using the same base folder for each run enables the reuse of the environment. Currently, each run creates the whole environment from scratch, which takes ~8mins on an M1 Mac. The refactored approach allows for the reuse of existing env, which skips the setup if the builds are up to date and move straight to the tests. Unfortunately, I don't think this is possible currently in the CI, but it helps a ton when working with the tests locally.


### ⚡︎ Run each branch setup in parallel

Fetching/building separate branches can be done in parallel, saving some extra time. CI efficiency is too erratic to give solid numbers, but testing locally, the refactored setup takes 50% less time than the current one when building from scratch. For my machine, which is an M1 Mac, it reduced from ~8 to ~4 minutes. 


### ⚡︎ Use the latest stable WordPress version instead of bleeding edge by default

We're currently using the WP bleeding edge (`core: "WordPress/WordPress"` in wp-env config) for the performance tests environment unless specified otherwise via the `--wp-version` option. I'm not sure if that's intentional, but it makes more sense to me to run performance tests against the latest production release instead (`core: null`) for more consistent measurements. 


## Testing Instructions

Checkout this branch (`refactor/perf-tests-tmp-folder`) and run `npm install`.
The following scenarios should be working as expected:

1. Compare the current branch with another branch or commit SHA. There should not be a need to pass the current branch's name, e.g.:

   ```
   bin/plugin/cli.js perf trunk
   ```

   ```
   bin/plugin/cli.js perf debd225d007f4e441ceec80fbd6fa96653f94737
   ```

2. Compare the current branch with two or more other git refs, e.g.:

   ```
   bin/plugin/cli.js perf trunk debd225d007f4e441ceec80fbd6fa96653f94737
   ```

3. Compare using a specific WP version:
 
   ```
   bin/plugin/cli.js perf debd225d007f4e441ceec80fbd6fa96653f94737 --wp-version 6.1
   ```

4. Rerun the above to ensure the setup is being reused:

   ```
   bin/plugin/cli.js perf debd225d007f4e441ceec80fbd6fa96653f94737 --wp-version 6.1
   ```

5. To ensure whether other specific comparisons work, e.g., the `Compare performance with current WordPress Core and previous Gutenberg versions` from the performance.yml workflow, you'll need to do the following.

   Assuming we're testing `release/15.6` vs. `release/15.5` and `wp/6.1`:
   - `git checkout release/15.6` as it's the one from which the workflow will be dispatched,
   - `git merge origin refactor/perf-test-results-path`,
   - `nvm use && npm install`,
   - `bin/plugin/cli.js perf release/15.5 wp/6.1 --wp-version 6.1`,
   - Confirm when asked to clear the environment,
   - Confirm when asked if the plugin should be built from the local branch.
   - Go make some coffee,
   - See the tests pass, 🤞 
   - `git reset --hard origin/release/15.6`.

6. The script should recognize if the current root branch name is passed explicitly and still use the local code, e.g.:

   ```
   bin/plugin/cli.js perf refactor/perf-test-results-path trunk
   ```

7. Finally, ensure that when comparing this branch with `trunk`, they should both show the same numbers. This should be proof that the new logic does not affect the perf measurements.

   ```
   bin/plugin/cli.js perf trunk
   ```
